### ISSUE TITLE
import private key crashes due to lack of permission

### ISSUE THREAD

**Issue by spaetz** [open-keychain/open-keychain#1733]
This is cyanogenmod 13 (android 6) on my moto g 2014. Trying to import my private key, I open the file manager, select my key file and BOOM OpenKeychain crashes. Stacktrace below indicates a lack of file permission, and the privacy settings do not show an file acces permission indeed.
This is OpenKeychain 3.8.2 from fdroid, in case this matters.

```
E/Keychain(27421): FileNotFoundException!
E/Keychain(27421): java.io.FileNotFoundException: /storage/emulated/0/spaetz/gpg_2016_sspaeth_65B22153_sec.asc: open failed: EACCES (Permission denied)
E/Keychain(27421): 	at libcore.io.IoBridge.open(IoBridge.java:452)
E/Keychain(27421): 	at java.io.FileInputStream.<init>(FileInputStream.java:76)
E/Keychain(27421): 	at java.io.FileInputStream.<init>(FileInputStream.java:103)
E/Keychain(27421): 	at android.content.ContentResolver.openInputStream(ContentResolver.java:662)
E/Keychain(27421): 	at org.sufficientlysecure.keychain.ui.adapter.ImportKeysListLoader.getInputData(ImportKeysListLoader.java:143)
E/Keychain(27421): 	at org.sufficientlysecure.keychain.ui.adapter.ImportKeysListLoader.loadInBackground(ImportKeysListLoader.java:76)
E/Keychain(27421): 	at org.sufficientlysecure.keychain.ui.adapter.ImportKeysListLoader.loadInBackground(ImportKeysListLoader.java:45)
E/Keychain(27421): 	at android.support.v4.content.AsyncTaskLoader.onLoadInBackground(AsyncTaskLoader.java:296)
E/Keychain(27421): 	at android.support.v4.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:54)
E/Keychain(27421): 	at android.support.v4.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:42)
E/Keychain(27421): 	at android.support.v4.content.ModernAsyncTask$2.call(ModernAsyncTask.java:128)
E/Keychain(27421): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/Keychain(27421): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
E/Keychain(27421): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
E/Keychain(27421): 	at java.lang.Thread.run(Thread.java:818)
E/Keychain(27421): Caused by: android.system.ErrnoException: open failed: EACCES (Permission denied)
E/Keychain(27421): 	at libcore.io.Posix.open(Native Method)
E/Keychain(27421): 	at libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)
E/Keychain(27421): 	at libcore.io.IoBridge.open(IoBridge.java:438)
E/Keychain(27421): 	... 14 more
E/AndroidRuntime(27421): FATAL EXCEPTION: ModernAsyncTask #1
E/AndroidRuntime(27421): Process: org.sufficientlysecure.keychain, PID: 27421
E/AndroidRuntime(27421): Theme: themes:{}
E/AndroidRuntime(27421): java.lang.RuntimeException: An error occurred while executing doInBackground()
E/AndroidRuntime(27421): 	at android.support.v4.content.ModernAsyncTask$3.done(ModernAsyncTask.java:142)
E/AndroidRuntime(27421): 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
E/AndroidRuntime(27421): 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
E/AndroidRuntime(27421): 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
E/AndroidRuntime(27421): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
E/AndroidRuntime(27421): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
E/AndroidRuntime(27421): 	at java.lang.Thread.run(Thread.java:818)
E/AndroidRuntime(27421): Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.io.InputStream org.sufficientlysecure.keychain.util.InputData.getInputStream()' on a null object reference
E/AndroidRuntime(27421): 	at org.sufficientlysecure.keychain.ui.adapter.ImportKeysListLoader.generateListOfKeyrings(ImportKeysListLoader.java:111)
E/AndroidRuntime(27421): 	at org.sufficientlysecure.keychain.ui.adapter.ImportKeysListLoader.loadInBackground(ImportKeysListLoader.java:77)
E/AndroidRuntime(27421): 	at org.sufficientlysecure.keychain.ui.adapter.ImportKeysListLoader.loadInBackground(ImportKeysListLoader.java:45)
E/AndroidRuntime(27421): 	at android.support.v4.content.AsyncTaskLoader.onLoadInBackground(AsyncTaskLoader.java:296)
E/AndroidRuntime(27421): 	at android.support.v4.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:54)
E/AndroidRuntime(27421): 	at android.support.v4.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:42)
E/AndroidRuntime(27421): 	at android.support.v4.content.ModernAsyncTask$2.call(ModernAsyncTask.java:128)
E/AndroidRuntime(27421): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/AndroidRuntime(27421): 	... 3 more
W/ActivityManager( 2987):   Force finishing activity org.sufficientlysecure.keychain/.ui.ImportKeysActivity
I/OpenGLRenderer( 2987): Initialized EGL, version 1.4
I/ThermalEngine( 2401): Sensor:batt_therm:27600 mC
I/Process (27421): Sending signal. PID: 27421 SIG: 9
I/WindowState( 2987): WIN DEATH: Window{45d67fe u0 org.sufficientlysecure.keychain/org.sufficientlysecure.keychain.ui.CreateKeyActivity}
D/GraphicsStats( 2987): Buffer count: 16
I/ActivityManager( 2987): Process org.sufficientlysecure.keychain (pid 27421) has died
W/ActivityManager( 2987): Force removing ActivityRecord{66e2cbd u0 org.sufficientlysecure.keychain/.ui.CreateKeyActivity t147}: app died, no saved state
```

**Comment by dschuermann** [comment]
Thanks for your report. Android 6 runtime permissions are requested in ImportKeysListFragment, but looks like there are some problems related to it. This issue needs to be repdouced and tested.

**Comment by spaetz** [comment]
Another bit of data. The private key file was saved on the sdcard which was formatted as "integrated" using androids 6's new feature. The file manager was cyanogenmods standard file manager.

I later openend the key file directly in the file manager anf chose "import with openkeychain" as action. It then asked me if I wanted to grant openkeychain permission to access files and photos.
In any case, I cannot reproduce this easily anymore, as it works now.

**Comment by runnerway** [comment]
Yes. I found the problem in the same way. 
I'm pretty sure this problem will disappear once the fix will be merged.

You can't reproduce it anymore because you have granted the permission now.

**Comment by petmue** [comment]
Got the same issue on Android 6, but this worked for me as well:

> Another bit of data. The private key file was saved on the sdcard which was formatted as "integrated" using androids 6's new feature. The file manager was cyanogenmods standard file manager.

> I later openend the key file directly in the file manager anf chose "import with openkeychain" as action. It then asked me if I wanted to grant openkeychain permission to access files and photos.
In any case, I cannot reproduce this easily anymore, as it works now.

**Comment by spaetz** [comment]
Final info, today a fix for cyanogenmods file manager was committed where it would crash as it failed to ask for necessary permissions. So it *might* have been a bug in the file manager after all. I have no idea whether this is relevant in this case at all, though.

**Comment by dschuermann** [comment]
Should be fixed in 3.9.1 and 3.9.2 thanks to @runnerway . Please report if it still exists.
